### PR TITLE
[RFC] Clang support

### PR DESCRIPTION
**Pull request by jforissier**
Supersedes #2944.

See usage information in the last patch ("Clang support").

### CODE REVIEW

**Comment by jforissier** [comment]
This is an RFC as a whole, but I think several commits could be merged regardless of Clang support. So feel free to provide R-b's, then I'll take those to a separate PR.

**Comment by jenswi-linaro** [comment]
For commits "build: print CC line when .am-defines.s is generated", "arm: remove -mthumb-interwork" and "core: add C compiler version to identification string":
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

For commit "core: change the pattern used to generate asm-defines.h":
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

File: lib/libutils/isoc/isalnum.c [hunk 1]
```diff
@@ -4,7 +4,7 @@
  */
 #include <ctype.h>
 
-int __builtin_isalnum(int c)
+int isalnum(int c)
```
Comment (jenswi-linaro) [inline lib/libutils/isoc/isalnum.c]: Comment in the commit "libutils: strip __builtin_ prefix from implementations":
How does this play with the define `isalnum()` in `<ctype.h>`?

**Comment by jenswi-linaro** [comment]
Comment on the commit "libutils: ctype.h: do not define functions as builtins"
In the GCC documentation it says:
*An OS kernel is an example of a program running in a freestanding environment*

OP-TEE is an OS kernel, so it's not obvious that `-ffreestanding` should be dropped. It seems that Linux doesn't use `-ffreestanding` unless required for certain files with Neon intrinsics so we're in good company dropping this.

By dropping this flag, did you notice any difference in the generated code?

File: core/arch/arm/sm/sm_a32.S [hunk 2]
```diff
@@ -20,32 +20,39 @@
...
 	/* User mode registers has to be saved from system mode */
 	cps	#CPSR_MODE_SYS
-	stm	r0!, {sp, lr}
+	str	sp, [r0], #4
+	str	lr, [r0], #4
 
 	cps	#CPSR_MODE_IRQ
 	mrs	r2, spsr
-	stm	r0!, {r2, sp, lr}
+	str_r2_sp_lr
```
Comment (jenswi-linaro) [inline core/arch/arm/sm/sm_a32.S]: I'd rather open code what's done in `str_r2_sp_lr` or define a macro that does all the stuff for a specified mode
Same goes for `ldr_r2_sp_lr` below.

**Comment by jenswi-linaro** [comment]
For commits "libutils: asm.S: remove .func/.endfunc directives"
and "core: thread_a32.S: use assembler unified syntax (UAL)"

`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

For commit "core: thread_a32.S: replace some ldm/stm with ldr/str"
Clarify that it's with ARMv8.2-LSMAOC that using these registers with
ldm/stm becomes optionally deprecated. With that:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

File: core/arch/arm/kernel/unwind_arm32.c [hunk 3]
```diff
@@ -476,11 +476,14 @@ vaddr_t *unw_get_kernel_stack(void)
 	size_t exidx_sz = 0;
 	vaddr_t *tmp = NULL;
 	vaddr_t *addr = NULL;
-	struct unwind_state_arm32 state = { 0 };
```
Comment (jenswi-linaro) [inline core/arch/arm/kernel/unwind_arm32.c]: The preferred struct initialization is now `{ }`

**Comment by jenswi-linaro** [comment]
For commit "libutils: setjmp_a32.S: split ldmfd and stmea containing sp":
There's a few more places that needs to be fixed. Please add a reference to ARMv8.2-LSMAOC in the commit message.

**Comment by jenswi-linaro** [comment]
Comment for "core, libutee: move .weak directive after the symbol definition"
It seems that a few places with the `.weak` directive are missed.

**Comment by jenswi-linaro** [comment]
Comment for "Force alignment of assembler functions (FUN and LOCAL_FUN) to 4 bytes":
It seems that a few places in particular in `_a64.S` files need to be updated too.

**Comment by jenswi-linaro** [comment]
Comment for "core: generic_entry_a32.S: remove SP from ldmdb/stmdb":
With a reference to ARMv8.2-LSMAOC:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by jenswi-linaro** [comment]
For commits "core: link.mk: generate empty tee-pageable.bin when pager is disabled" and "core: arm: thread.c: increase stack_tmp size from 1.5 to to 2KiB"

Acked-by: Jens Wiklander <jens.wiklander@linaro.org>

**Comment by jenswi-linaro** [comment]
Comments for commit "core: fix Clang warning":
The first line is a bit unspecific, with that fixed:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

File: core/arch/arm/include/arm64.h [hunk 4]
```diff
@@ -258,19 +258,23 @@ static inline void tlbi_vale1is(uint64_t mva)
...
-static inline type read_##reg(void)			\
-{							\
-	type val;					\
-							\
-	asm volatile("mrs %0, " #asmreg : "=r" (val));	\
-	return val;					\
+#define DEFINE_REG_READ_FUNC_(reg, type, asmreg)		\
+static inline type read_##reg(void)				\
+{								\
+	uint64_t val64;						\
```
Comment (jenswi-linaro) [inline core/arch/arm/include/arm64.h]: Please initialize.

File: core/arch/arm/include/arm64.h [hunk 5]
```diff
@@ -258,19 +258,23 @@ static inline void tlbi_vale1is(uint64_t mva)
...
+	type val;						\
+								\
+	asm volatile("mrs %0, " #asmreg : "=r" (val64));	\
+	val = (type)val64;					\
```
Comment (jenswi-linaro) [inline core/arch/arm/include/arm64.h]: Doesn't a plain `return val64;` work?

File: ta/arch/arm/user_ta_header.c [hunk 6]
```diff
@@ -25,26 +25,44 @@ const char trace_ext_prefix[]  = "TA";
 /* exprted to user_ta_header.c, built within TA */
 struct utee_params;
 
+#if defined(ARM32) && defined(CFG_UNWIND)
```
Comment (jenswi-linaro) [inline ta/arch/arm/user_ta_header.c]: This is doesn't look very nice, how about moving it into an `_a32.S` file?

File: lib/libutils/isoc/newlib/bcmp.c [hunk 7]
```diff
@@ -0,0 +1,41 @@
+/*
```
Comment (jenswi-linaro) [inline lib/libutils/isoc/newlib/bcmp.c]: SPDX and license header?

**Comment by jenswi-linaro** [comment]
Comment for the commit "core: pager: arm32: remove .ARM.exidx section from tee_pager.bin"
By moving `.ARM.exidx` to `tee-pageable.bin` we can't do reliable stack tracing from any unpaged mode (typically abort mode). Or am I missing something?

**Comment by jenswi-linaro** [comment]
For commits "ta: ta.ld.S: move .dynamic section next to .got", 
"ta: ldelf.ld.S: move .dynamic section next to .got",
"ldelf: scripts/gen_ldelf_hex.py: allow multiple RW load segments" and
"libutils/newlib: sort sub.mk in alphabetical order":
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

For the commit "core_self_tests.c: build with -fno-builtin":
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by jenswi-linaro** [comment]
How about avoiding the shortcomings of `llvm-objcopy` by simply adding a python script that dumps out all the needed parts from the `tee.elf` files? Some external builds still depends on some of the artifact `tee-pager.bin` and possibly some other file but I guess we could create those too if needed.

File: lib/libutils/isoc/isalnum.c (hunk 1 above)
Comment (jforissier) [reply lib/libutils/isoc/isalnum.c]: Well, due to the define, the commit "libutils: strip _builtin prefix from implementations" is effectively a no-op... just a preparation for the following commit actually ("libutils: ctype.h: do not define functions as builtins"). I guess I'd better squash them.

To answer your question below:

> By dropping this flag [-ffreestanding], did you notice any difference in the generated code?

No. The code is totally identical. I compared `tee.dmp` with and without the two libutils patches applied. The only differences I see in tee.dmp are a few symbols are named differently (`__func__.6169` becomes `__func__.6195` etc.).

It is possible however to see one effect of `-ffreestanding` on some made-up example:

```
$ cat test.c
#include <stdio.h>
int main(int argc, char *argv[])
{
        printf("A");
        return 0;
}
$ arm-linux-gnueabihf-gcc -c test.c -o nofreestanding.o
$ arm-linux-gnueabihf-gcc -c test.c -ffreestanding -o freestanding.o
$ arm-linux-gnueabihf-nm freestanding.o | grep -E "(printf|putchar)"
         U printf
$ arm-linux-gnueabihf-nm nofreestanding.o | grep -E "(printf|putchar)"
         U putchar
```
The flag prevented the printf -> putchar optimization. So, better avoid it if we can.

File: core/arch/arm/sm/sm_a32.S (hunk 2 above)
Comment (jforissier) [reply core/arch/arm/sm/sm_a32.S]: I'll introduce macros (`save_regs mode` and `restore_regs mode`).

File: core/arch/arm/kernel/unwind_arm32.c (hunk 3 above)
Comment (jforissier) [reply core/arch/arm/kernel/unwind_arm32.c]: Will fix

**Comment by jforissier** [comment]
> For commit "libutils: setjmp_a32.S: split ldmfd and stmea containing sp":
> There's a few more places that needs to be fixed.

Where? I can only find `pm_a32.S` that I have missed:

```
$ git grep -E -i 'ldm.*\{.*sp.*\}'
core/arch/arm/sm/pm_a32.S:      ldmia   r0!, {sp, pc}
$ git grep -E -i 'stm.*\{.*(sp|pc).*\}'
$
```

> Please add a reference to ARMv8.2-LSMAOC in the commit message.

I will.

**Comment by jforissier** [comment]
> Comment for "core, libutee: move .weak directive after the symbol definition"
> It seems that a few places with the .weak directive are missed.

Found one more occurrence (`plat_cpu_reset_early`), now fixed.

**Comment by jforissier** [comment]
>Comment for "Force alignment of assembler functions (FUN and LOCAL_FUN) to 4 bytes":
It seems that a few places in particular in _a64.S files need to be updated too.

Where?

File: core/arch/arm/include/arm64.h (hunk 5 above)
Comment (jforissier) [reply core/arch/arm/include/arm64.h]: It does. Will fix.

File: core/arch/arm/include/arm64.h (hunk 4 above)
Comment (jforissier) [reply core/arch/arm/include/arm64.h]: OK

File: lib/libutils/isoc/newlib/bcmp.c (hunk 7 above)
Comment (jforissier) [reply lib/libutils/isoc/newlib/bcmp.c]: There is no license and no copyright in the [upstream file](https://www.sourceware.org/git/gitweb.cgi?p=newlib-cygwin.git;a=blob;f=newlib/libc/string/bcmp.c;h=747f8af51d0d164fb2b5e01a89ab169ec2ac7df6;hb=HEAD) I've used.
Instead there is a big [`COPYING.NEWLIB`](https://www.sourceware.org/git/gitweb.cgi?p=newlib-cygwin.git;a=blob;f=COPYING.NEWLIB;h=0c39f9031c1efbb4daf8dbb4e6324b77edd4b2f2;hb=HEAD) file. So it seems to attribute the proper copyright and license to `bcmp.c`.

Why is it that most of our files from newlib have a BSD-3-Clause license and RedHat copyright? Coming from a different repository perhaps?

**Comment by jforissier** [comment]
> Comment for the commit "core: pager: arm32: remove .ARM.exidx section from tee_pager.bin"
> By moving .ARM.exidx to tee-pageable.bin we can't do reliable stack tracing from any unpaged mode (typically abort mode). Or am I missing something?

I think you're right :-/
I was mislead by the linker, which rightly warns about the situation but our use case is quite special -- we know the sections will be available at runtime in the pager binary. I'll find a way to silence the warning and comment why it is the right approach.

**Comment by jenswi-linaro** [comment]
>>Comment for "Force alignment of assembler functions (FUN and LOCAL_FUN) to 4 bytes":
It seems that a few places in particular in _a64.S files need to be updated too.

> Where?

I was thinking of
https://github.com/jforissier/optee_os/blob/30d524bbb850b8f86695e59d26177bec5856b7ae/core/arch/arm/kernel/thread_a64.S#L413
but perhaps that's not needed

**Comment by jenswi-linaro** [comment]
>> For commit "libutils: setjmp_a32.S: split ldmfd and stmea containing sp":
There's a few more places that needs to be fixed.

> Where? I can only find pm_a32.S that I have missed...

I think you're right, must have been searching in some other tree.

**Comment by jforissier** [comment]
Regarding code alignment, I think my FUNC_ALIGN macro is not needed because it is always possible to increase alignment to larger than 4 by adding `.align` before `FUNC`. I'll rework this.

**Comment by jforissier** [comment]
Regarding ldm/stm:
- Does the deprecation really has something to do with ARMv8.2-LSMAOC?
- The constraints are very specific and it seems we're OK. See [here](https://github.com/llvm/llvm-project/blame/0fa832002c27895ce4d9f329ce5b6205abc22165/llvm/test/MC/ARM/arm-load-store-multiple-deprecated.s).

File: lib/libutils/isoc/newlib/bcmp.c (hunk 7 above)
Comment (jenswi-linaro) [reply lib/libutils/isoc/newlib/bcmp.c]: No, it's from the same. We asked someone at Redhat for clarification at some point and as a service to those using OP-TEE we're adding the license to make it easier to FOSS in the file.

**Comment by jenswi-linaro** [comment]
> Does the deprecation really has something to do with ARMv8.2-LSMAOC?

Yes, that's the only reason I could find that makes sense. If we don't care about ARMv8.2-LSMAOC then we should just silent the warnings by some means instead of making changes to the code.

**Comment by jforissier** [comment]
The LDM/STM deprecations date back to ARMv7 as mentioned in this LLVM patch: http://lists.llvm.org/pipermail/llvm-commits/Week-of-Mon-20141215/249704.html ("ARM ARM").

IMO we should rewrite the code, but I'd rather do it in a single commit which clearly state what is deprecated.

**Comment by jenswi-linaro** [comment]
> IMO we should rewrite the code, but I'd rather do it in a single commit which clearly state what is deprecated.

Sounds good to me.

File: lib/libutils/isoc/newlib/bcmp.c (hunk 7 above)
Comment (jforissier) [reply lib/libutils/isoc/newlib/bcmp.c]: OK, I will add a SPDX line, BSD-2-Clause is certainly applicable.
As for the copyright mention, how do we know which one in COPYING.NEWLIB applies to bcmp.c? I mean, we can't be sure RedHat wrote this. Perhaps it should be attributed to someone else. And therefore, I think we'd rather not add anything else to the file. Just the SPDX line.

File: ta/arch/arm/user_ta_header.c (hunk 6 above)
Comment (jforissier) [reply ta/arch/arm/user_ta_header.c]: Agreed. I will clean this up.

**Comment by jforissier** [comment]
Rebased and updated in a new PR.

**Pull request closed without merging**